android-sdk/README.md

```diff
@@ -0,0 +1,224 @@
+        this.subjectTokenType = config.getProperty("subject_token_type");
+        this.requestedTokenType = config.getProperty("requested_token_type");
+        this.realm = config.getProperty("realm");
+        this.core = core;
+    }
+}
+```
+
+#### Registering a Service Module and the ServiceModuleRegistry
+The Core SDK will provide a service registry that modules can register themselves with as well as register their dependencies.  During build MobileCore will call bootstrap on services in the order of those with the least dependencies to those with the most dependencies.
+
+Example registration of KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+  static {
+    ServiceModuleRegistry.registerServiceModule("keycloak", KeyCloakService.class, "http");
+  }
+  /*snip*/
+}
+```
+
+ServiceModuleRegistry.java
+```
+
+/**
+ * Registry of service modules.  It is responsible for maintaining a mapping of service modules to
+ * their types as well as the dependencies each module expects.
+ */
+public final class ServiceModuleRegistry {
+    private static final Map<String, Class<? extends ServiceModule>> serviceTypeMap = new HashMap<>();
+    private static final Map<String, String> dependencyMap = new HashMap<>();
+    /**
+     * Register the looking for a service module to a class.
+     *
+     * @param type the type of module to use
+     * @param moduleClass the class that implements a service module type
+     * @param dependsOn the other serviceTypes this Service Depends on.
+     */
+    public static void registerServiceModule(String type, Class<? extends ServiceModule> moduleClass, String... dependsOn) {
+        serviceTypeMap.put(type, moduleClass);
+        for (String dependency : dependsOn) {
+            dependencyMap.put(type, dependency);
+        }
+    }
+
+}
+```
+** Discussion Point ** I'm not a fan of static methods during testing.  Perhaps there should be a method in the ServiceModule interface which returns a list of services that the service depends on.  MobileCore could then read that list by invoking that method as part of bootstrapping.
+
+#### Referencing a Service
+An instance of the core bassed during bootstrap can be used to reference other services.  MobileCore is responsible for ensuring that ServiceModule.bootstrap is not called until all declared dependencies are bootstrapped.
+
+SyncService.java
+```
+public class SyncService implements ServiceModule {
+
+  KeyCloakService keycloakService;
+  /*snip*/
+  public void bootstrap(MobileCore core, ServiceConfiguration config) {
+     this.keycloakService = core.getService("keycloak", KeyCloakService.class);
+     this.networkService = core.getService("network", NetworkService.class);
+     /*snip*/
+  }
+
+  public void performSync() {
+    if (keycloakService.isLoggedIn()) {
+      Reuqest request = networkService.newRequest();
+      request.addSecurityHeaders(keycloakService.getHeaders());//pseudocode method, don't expect this exactly.
```

Comment:
-1 for this aproach. If I will have ability to change single element in proposal it will be this statement.
Imposing hard dependencies on the services itself (managed by us) it's not flexible aproach and it may be hard to maintain. It has numerous implications that server layer needs to handle. 

Happy to explain that carefully 🗣 

TL;DR - That's imposing hard dependency on sync service and it's less flexible.

Reply:
What approach exactly do you mean @wtrocki ?

Reply:
Sync have reference to keycloak service.

Reply:
Looks like I missed this stuff, but strongly agree with @wtrocki here. We don't want to place hard dependencies like this IMO